File: bmmpy/ga.py
"""
Genetic-algorithm optimizers for bmmpy.
"""

from __future__ import annotations

from collections.abc import Sequence

from ._bmmpy import (
    Candidate,
    GeneticAlgorithm as _NativeGeneticAlgorithm,
    GeneticAlgorithmConfig,
    IslandModel as _NativeIslandModel,
    IslandModelConfig,
    IslandModelSnapshot,
    IslandSnapshot,
    IslandSpec,
    MigrationPolicy,
    OverflowPolicy as _NativeOverflowPolicy,
    RowWindow,
    RunStats,
    StopCriteria,
)

OverflowPolicy = _NativeOverflowPolicy

StopCriteria.__doc__ = "Stopping rules for the genetic optimizers."
RunStats.__doc__ = "Runtime counters collected during a genetic run."
MigrationPolicy.__doc__ = "Per-island migration settings."
IslandSpec.__doc__ = "Static configuration for one island in the island model."
IslandSnapshot.__doc__ = "Per-island runtime snapshot."
IslandModelSnapshot.__doc__ = "Whole-model runtime snapshot."
GeneticAlgorithmConfig.__doc__ = "Low-level native GA configuration."
IslandModelConfig.__doc__ = "Low-level native island-model configuration."


def _build_stop_criteria(
    *,
    max_generations: int | None,
    max_stale_generations: int | None,
    target_total_weight: int | None,
) -> StopCriteria:
    stop = StopCriteria()
    stop.max_generations = max_generations
    stop.max_stale_generations = max_stale_generations
    stop.target_total_weight = target_total_weight
    return stop


def _clone_stop_criteria(stop: StopCriteria) -> StopCriteria:
    clone = StopCriteria()
    clone.max_generations = stop.max_generations
    clone.max_stale_generations = stop.max_stale_generations
    clone.target_total_weight = stop.target_total_weight
    return clone


def _build_ga_config(
    *,
    population_size: int,
    elite_count: int,
    tournament_size: int,
    mutation_rate: float,
    max_generations: int | None,
    max_stale_generations: int | None,
    target_total_weight: int | None,
    seed: int,
) -> GeneticAlgorithmConfig:
    config = GeneticAlgorithmConfig()
    config.population_size = population_size
    config.elite_count = elite_count
    config.tournament_size = tournament_size
    config.mutation_rate = mutation_rate
    config.stop = _build_stop_criteria(
        max_generations=max_generations,
        max_stale_generations=max_stale_generations,
        target_total_weight=target_total_weight,
    )
    config.seed = seed
    return config


def _clone_ga_config(config: GeneticAlgorithmConfig) -> GeneticAlgorithmConfig:
    clone = GeneticAlgorithmConfig()
    clone.population_size = config.population_size
    clone.elite_count = config.elite_count
    clone.tournament_size = config.tournament_size
    clone.mutation_rate = config.mutation_rate
    clone.stop = _clone_stop_criteria(config.stop)
    clone.seed = config.seed
    return clone


def _resolve_overflow_policy(
    policy: str | OverflowPolicy,
) -> OverflowPolicy:
    if isinstance(policy, str):
        normalized = policy.strip().lower()
        if normalized == "drop_oldest":
            return OverflowPolicy.DropOldest
        if normalized == "drop_newest":
            return OverflowPolicy.DropNewest
        if normalized == "reject_publish":
            return OverflowPolicy.RejectPublish
        raise ValueError(f"Unsupported overflow_policy: {policy!r}")

    if policy not in (
        OverflowPolicy.DropOldest,
        OverflowPolicy.DropNewest,
        OverflowPolicy.RejectPublish,
    ):
        raise ValueError(f"Unsupported overflow_policy: {policy!r}")

    return policy


def _common_ga_kwargs_are_default(
    *,
    population_size: int,
    elite_count: int,
    tournament_size: int,
    mutation_rate: float,
    max_generations: int | None,
    max_stale_generations: int | None,
    target_total_weight: int | None,
    seed: int,
    seed_stride: int,
) -> bool:
    return (
        population_size == 300
        and elite_count == 3
        and tournament_size == 3
        and mutation_rate == 0.3
        and max_generations is None
        and max_stale_generations is None
        and target_total_weight is None
        and seed == 0
        and seed_stride == 1
    )


class GeneticAlgorithm:
    """
    Genetic optimizer over a row window.

    The returned value is a full list of Candidate objects with length equal to
    the input window size. Each candidate describes one output row of the found
    transform.
    """

    __slots__ = (
        "population_size",
        "elite_count",
        "tournament_size",
        "mutation_rate",
        "max_generations",
        "max_stale_generations",
        "target_total_weight",
        "seed",
        "_impl",
    )

    def __init__(
        self,
        *,
        population_size: int = 300,
        elite_count: int = 3,
        tournament_size: int = 3,
        mutation_rate: float = 0.3,
        max_generations: int | None = None,
        max_stale_generations: int | None = None,
        target_total_weight: int | None = None,
        seed: int = 0,
    ) -> None:
        self.population_size = population_size
        self.elite_count = elite_count
        self.tournament_size = tournament_size
        self.mutation_rate = mutation_rate
        self.max_generations = max_generations
        self.max_stale_generations = max_stale_generations
        self.target_total_weight = target_total_weight
        self.seed = seed

        config = _build_ga_config(
            population_size=population_size,
            elite_count=elite_count,
            tournament_size=tournament_size,
            mutation_rate=mutation_rate,
            max_generations=max_generations,
            max_stale_generations=max_stale_generations,
            target_total_weight=target_total_weight,
            seed=seed,
        )
        self._impl = _NativeGeneticAlgorithm(config)

    def __repr__(self) -> str:
        return (
            "GeneticAlgorithm("
            f"population_size={self.population_size}, "
            f"elite_count={self.elite_count}, "
            f"tournament_size={self.tournament_size}, "
            f"mutation_rate={self.mutation_rate}, "
            f"max_generations={self.max_generations}, "
            f"max_stale_generations={self.max_stale_generations}, "
            f"target_total_weight={self.target_total_weight}, "
            f"seed={self.seed})"
        )

    def name(self) -> str:
        return self._impl.name()

    def initialize(self, window: RowWindow) -> None:
        self._impl.initialize(window)

    def step(self) -> None:
        self._impl.step()

    def done(self) -> bool:
        return self._impl.done()

    def generation(self) -> int:
        return self._impl.generation()

    def best_individual(self) -> list[Candidate]:
        return self._impl.best_individual()

    def stats(self) -> RunStats:
        return self._impl.stats()

    def best_score(self) -> int:
        return self._impl.best_score()

    def optimize(self, window: RowWindow) -> list[Candidate]:
        return self._impl.optimize(window)


class IslandModel:
    """
    Island-model wrapper around GeneticAlgorithm.

    By default this class builds one GA configuration per island from the common
    keyword arguments. For heterogeneous islands you can pass ga_configs with
    one GeneticAlgorithmConfig per IslandSpec.
    """

    __slots__ = (
        "island_count",
        "channel_capacity",
        "overflow_policy",
        "_impl",
    )

    def __init__(
        self,
        *,
        islands: Sequence[IslandSpec],
        ga_configs: Sequence[GeneticAlgorithmConfig] | None = None,
        population_size: int = 300,
        elite_count: int = 3,
        tournament_size: int = 3,
        mutation_rate: float = 0.3,
        max_generations: int | None = None,
        max_stale_generations: int | None = None,
        target_total_weight: int | None = None,
        seed: int = 0,
        seed_stride: int = 1,
        channel_capacity: int = 256,
        overflow_policy: str | OverflowPolicy = "drop_oldest",
    ) -> None:
        island_list = list(islands)
        if not island_list:
            raise ValueError("IslandModel requires at least one island")
        if channel_capacity <= 0:
            raise ValueError("channel_capacity must be greater than zero")
        if seed_stride < 0:
            raise ValueError("seed_stride must be non-negative")

        if ga_configs is not None:
            if len(ga_configs) != len(island_list):
                raise ValueError(
                    "ga_configs length must match the number of islands"
                )
            if not _common_ga_kwargs_are_default(
                population_size=population_size,
                elite_count=elite_count,
                tournament_size=tournament_size,
                mutation_rate=mutation_rate,
                max_generations=max_generations,
                max_stale_generations=max_stale_generations,
                target_total_weight=target_total_weight,
                seed=seed,
                seed_stride=seed_stride,
            ):
                raise ValueError(
                    "Pass either ga_configs or common GA keyword arguments, not both"
                )
            normalized_ga_configs = [_clone_ga_config(config) for config in ga_configs]
        else:
            normalized_ga_configs = [
                _build_ga_config(
                    population_size=population_size,
                    elite_count=elite_count,
                    tournament_size=tournament_size,
                    mutation_rate=mutation_rate,
                    max_generations=max_generations,
                    max_stale_generations=max_stale_generations,
                    target_total_weight=target_total_weight,
                    seed=seed + index * seed_stride,
                )
                for index in range(len(island_list))
            ]

        config = IslandModelConfig()
        config.islands = island_list

        resolved_overflow_policy = _resolve_overflow_policy(overflow_policy)

        self.island_count = len(island_list)
        self.channel_capacity = channel_capacity
        self.overflow_policy = resolved_overflow_policy
        self._impl = _NativeIslandModel(
            config,
            normalized_ga_configs,
            channel_capacity,
            resolved_overflow_policy,
        )

    def __repr__(self) -> str:
        return (
            "IslandModel("
            f"island_count={self.island_count}, "
            f"channel_capacity={self.channel_capacity}, "
            f"overflow_policy={self.overflow_policy})"
        )

    def name(self) -> str:
        return self._impl.name()

    def initialize(self, window: RowWindow) -> None:
        self._impl.initialize(window)

    def start(self) -> None:
        self._impl.start()

    def request_stop(self) -> None:
        self._impl.request_stop()

    def wait(self) -> None:
        self._impl.wait()

    def running(self) -> bool:
        return self._impl.running()

    def stop_requested(self) -> bool:
        return self._impl.stop_requested()

    def best_individual(self) -> list[Candidate]:
        return self._impl.best_individual()

    def best_score(self) -> int:
        return self.snapshot().best_score

    def snapshot(self) -> IslandModelSnapshot:
        return self._impl.snapshot()

    def run_to_completion(self, window: RowWindow) -> list[Candidate]:
        return self._impl.run_to_completion(window)


__all__ = [
    "GeneticAlgorithm",
    "GeneticAlgorithmConfig",
    "IslandModel",
    "IslandModelConfig",
    "IslandModelSnapshot",
    "IslandSnapshot",
    "IslandSpec",
    "MigrationPolicy",
    "OverflowPolicy",
    "RunStats",
    "StopCriteria",
]

File: python/bindings/ga.cpp
#include "bindings.hpp"
#include "bmmpy/ga/genetic_algorithm.hpp"
#include "bmmpy/ga/island_model.hpp"
#include "bmmpy/ga/migration/bounded_channel.hpp"

#include <algorithm>
#include <cstddef>
#include <memory>
#include <nanobind/stl/optional.h>
#include <nanobind/stl/vector.h>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace nb = nanobind;

namespace {
namespace ga = ::bmmpy::ga;
namespace migration = ::bmmpy::ga::migration;

ga::IslandModel make_island_model(ga::IslandModelConfig config,
                                  std::vector<ga::GeneticAlgorithmConfig> ga_configs,
                                  std::size_t channel_capacity,
                                  migration::OverflowPolicy overflow_policy) {
    if (config.islands.empty()) {
        throw std::invalid_argument("IslandModel: islands must not be empty");
    }

    if (ga_configs.size() != config.islands.size()) {
        throw std::invalid_argument("IslandModel: ga_configs size must match config.islands size");
    }

    if (channel_capacity == 0) {
        throw std::invalid_argument("IslandModel: channel_capacity must be greater than zero");
    }

    std::vector<std::pair<std::size_t, ga::GeneticAlgorithmConfig>> configs_by_island;
    configs_by_island.reserve(config.islands.size());

    for (std::size_t index = 0; index < config.islands.size(); ++index) {
        const std::size_t island_id = config.islands[index].island_id;
        const bool duplicate_id =
            std::any_of(configs_by_island.begin(),
                        configs_by_island.end(),
                        [island_id](const auto& entry) { return entry.first == island_id; });

        if (duplicate_id) {
            throw std::invalid_argument("IslandModel: island_id values must be unique");
        }

        configs_by_island.emplace_back(island_id, ga_configs[index]);
    }

    ga::AlgorithmFactory factory =
        [configs_by_island = std::move(configs_by_island)](
            const ga::IslandSpec& spec) -> std::unique_ptr<ga::Algorithm> {
        const auto it =
            std::find_if(configs_by_island.begin(),
                         configs_by_island.end(),
                         [&spec](const auto& entry) { return entry.first == spec.island_id; });

        if (it == configs_by_island.end()) {
            throw std::invalid_argument(
                "IslandModel: missing GeneticAlgorithmConfig for island_id=" +
                std::to_string(spec.island_id));
        }

        return std::make_unique<ga::GeneticAlgorithm>(it->second);
    };

    auto channel = std::make_unique<migration::BoundedChannel>(channel_capacity, overflow_policy);

    return ga::IslandModel(std::move(config), std::move(factory), std::move(channel));
}

} // namespace

namespace bmmpy::bindings {

void bind_ga(nb::module_& m) {
    nb::enum_<migration::OverflowPolicy>(m, "OverflowPolicy")
        .value("DropOldest", migration::OverflowPolicy::DropOldest)
        .value("DropNewest", migration::OverflowPolicy::DropNewest)
        .value("RejectPublish", migration::OverflowPolicy::RejectPublish);

    nb::class_<ga::StopCriteria>(m, "StopCriteria")
        .def(nb::init<>())
        .def_rw("max_generations", &ga::StopCriteria::max_generations)
        .def_rw("max_stale_generations", &ga::StopCriteria::max_stale_generations)
        .def_rw("target_total_weight", &ga::StopCriteria::target_total_weight);

    nb::class_<ga::RunStats>(m, "RunStats")
        .def(nb::init<>())
        .def_rw("generations", &ga::RunStats::generations)
        .def_rw("evaluations", &ga::RunStats::evaluations)
        .def_rw("stale_generations", &ga::RunStats::stale_generations)
        .def_rw("seed", &ga::RunStats::seed);

    nb::class_<ga::MigrationPolicy>(m, "MigrationPolicy")
        .def(nb::init<>())
        .def_rw("interval_generations", &ga::MigrationPolicy::interval_generations)
        .def_rw("export_count", &ga::MigrationPolicy::export_count)
        .def_rw("import_count", &ga::MigrationPolicy::import_count)
        .def_rw("shared_pool_capacity", &ga::MigrationPolicy::shared_pool_capacity);

    nb::class_<ga::IslandSpec>(m, "IslandSpec")
        .def(nb::init<>())
        .def_rw("island_id", &ga::IslandSpec::island_id)
        .def_rw("migration", &ga::IslandSpec::migration);

    nb::class_<ga::IslandSnapshot>(m, "IslandSnapshot")
        .def(nb::init<>())
        .def_rw("island_id", &ga::IslandSnapshot::island_id)
        .def_rw("running", &ga::IslandSnapshot::running)
        .def_rw("stop_requested", &ga::IslandSnapshot::stop_requested)
        .def_rw("finished", &ga::IslandSnapshot::finished)
        .def_rw("stats", &ga::IslandSnapshot::stats)
        .def_rw("best_score", &ga::IslandSnapshot::best_score)
        .def_rw("best_individual", &ga::IslandSnapshot::best_individual);

    nb::class_<ga::IslandModelSnapshot>(m, "IslandModelSnapshot")
        .def(nb::init<>())
        .def_rw("running", &ga::IslandModelSnapshot::running)
        .def_rw("stop_requested", &ga::IslandModelSnapshot::stop_requested)
        .def_rw("island_count", &ga::IslandModelSnapshot::island_count)
        .def_rw("total_generations", &ga::IslandModelSnapshot::total_generations)
        .def_rw("best_score", &ga::IslandModelSnapshot::best_score)
        .def_rw("best_individual", &ga::IslandModelSnapshot::best_individual)
        .def_rw("islands", &ga::IslandModelSnapshot::islands);

    nb::class_<ga::GeneticAlgorithmConfig>(m, "GeneticAlgorithmConfig")
        .def(nb::init<>())
        .def_rw("population_size", &ga::GeneticAlgorithmConfig::population_size)
        .def_rw("elite_count", &ga::GeneticAlgorithmConfig::elite_count)
        .def_rw("tournament_size", &ga::GeneticAlgorithmConfig::tournament_size)
        .def_rw("num_parents", &ga::GeneticAlgorithmConfig::num_parents)
        .def_rw("num_offspring", &ga::GeneticAlgorithmConfig::num_offspring)
        .def_rw("enable_catastrophe", &ga::GeneticAlgorithmConfig::enable_catastrophe)
        .def_rw("catastrophe_threshold", &ga::GeneticAlgorithmConfig::catastrophe_threshold)
        .def_rw("catastrophe_survival_rate", &ga::GeneticAlgorithmConfig::catastrophe_survival_rate)
        .def_rw("mutation_rate", &ga::GeneticAlgorithmConfig::mutation_rate)
        .def_rw("stop", &ga::GeneticAlgorithmConfig::stop)
        .def_rw("seed", &ga::GeneticAlgorithmConfig::seed);

    nb::class_<ga::IslandModelConfig>(m, "IslandModelConfig")
        .def(nb::init<>())
        .def_rw("islands", &ga::IslandModelConfig::islands);

    nb::class_<ga::GeneticAlgorithm>(m, "GeneticAlgorithm")
        .def(nb::init<ga::GeneticAlgorithmConfig>(),
             nb::arg("config") = ga::GeneticAlgorithmConfig{})
        .def("name", &ga::GeneticAlgorithm::name)
        .def("initialize",
             &ga::GeneticAlgorithm::initialize,
             nb::arg("window"),
             nb::keep_alive<1, 2>())
        .def("step", &ga::GeneticAlgorithm::step, nb::call_guard<nb::gil_scoped_release>())
        .def("done", &ga::GeneticAlgorithm::done)
        .def("generation", &ga::GeneticAlgorithm::generation)
        .def("best_individual", &ga::GeneticAlgorithm::best_individual)
        .def("stats", &ga::GeneticAlgorithm::stats)
        .def("best_score", &ga::GeneticAlgorithm::best_score)
        .def("optimize",
             &ga::GeneticAlgorithm::optimize,
             nb::arg("window"),
             nb::call_guard<nb::gil_scoped_release>());

    nb::class_<ga::IslandModel>(m, "IslandModel")
        .def(
            "__init__",
            [](ga::IslandModel* self,
               ga::IslandModelConfig config,
               std::vector<ga::GeneticAlgorithmConfig> ga_configs,
               std::size_t channel_capacity,
               migration::OverflowPolicy overflow_policy) {
                new (self) ga::IslandModel(make_island_model(
                    std::move(config), std::move(ga_configs), channel_capacity, overflow_policy));
            },
            nb::arg("config"),
            nb::arg("ga_configs"),
            nb::arg("channel_capacity") = std::size_t{256},
            nb::arg("overflow_policy") = migration::OverflowPolicy::DropOldest)
        .def("name", &ga::IslandModel::name)
        .def("initialize", &ga::IslandModel::initialize, nb::arg("window"))
        .def("start", &ga::IslandModel::start)
        .def("request_stop", &ga::IslandModel::request_stop)
        .def("wait", &ga::IslandModel::wait, nb::call_guard<nb::gil_scoped_release>())
        .def("running", &ga::IslandModel::running)
        .def("stop_requested", &ga::IslandModel::stop_requested)
        .def("best_individual", &ga::IslandModel::best_individual)
        .def("snapshot", &ga::IslandModel::snapshot)
        .def("run_to_completion",
             &ga::IslandModel::run_to_completion,
             nb::arg("window"),
             nb::call_guard<nb::gil_scoped_release>());
}

} // namespace bmmpy::bindings